docs/architecture.md

```diff
@@ -0,0 +1,171 @@
+
+The w3filecoin pipeline is modeled into 4 different SST Stacks that will be have their infrastructure provisioned in AWS via AWS CloudFormation. These are:
+
+- API Stack
+- Consumer Stack
+- DB Stack
+- CRON Stack
+
+![Architecture](./architecture.png)
+
+## API Stack
+
+TODO
+API Gateway to expose:
+- Report API for failed aggregates to land into Storage Providers
+- Get to know state of deals
+- ...
+
+## Consumer Stack
+
+w3filecoin relies on R2 bucket HTTP URLs as the source that Storage Providers will use to fetch CAR files. This way, we need to wait on the replicator to write CAR files into R2. In addition, w3filecoin requires that a `pieceCid` and `pieceSize` is computed for 
+
+Once the above conditions are met, an external source triggers an event to the filecoin pipeline to let it know of a new car to be aggregated. Note that this allows us to make filecoin pipeline work with both w3up and CF Hosted APIs.
+
+Further down the line, in case we make writing to R2 first to better optimize gateway performance, it will also be an easy path forward to hook Filecoin pipeline with it, by leveraging receipts.
+
+For MVP the w3infra replicator lambda sends event to the event bridge, being `w3filecoin` responsible for listening on the event bridge for `car-replicated` events.
+
+## DB Stack
+
+To get CAR files into an aggregated deal, we will need to persist the metadata of these files pending being added to a Filecoin deal in a queue data structure. Potentially, this data structure could also have the concept of priority.
+
+Queue consumers can be triggered (e.g. via a CRON job) to grab items from the Queue and attempt to create an aggregate with them. In case an aggregate can be created, their state should be modified. We can see this as loading up a ferry with cargo once this shipment is ready to being offered via a broker.
+
+### Schema
+
+There are two tables within the DB Stack:
+* Cargo - keeps track of cargo (CAR Files ingested by web3.storage) state
+* Ferry - keeps track of aggregates state
+
+Both tables enable an implementation of a Priority Queued based on a State Machine on top of AWS RDS.
+
+```sql
+CREATE TABLE cargo
+(
+	-- Filecoin Piece CID - commP of CAR file
+	link TEXT PRIMARY KEY,
+	-- Filecoin Piece Size
+  size number NOT NULL,
+	-- CAR file CID
+	car_link TEXT NOT NULL,
+  -- State of the cargo in the pipeline
+  state CARGO_STATE NOT NULL,
+  -- Priority in the queue - for now likely same as queuedAt
+  priority TEXT NOT NULL,
+  -- Timestamp
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
+  -- TODO: Maybe timestamps for other stats?
+  -- Filecoin Aggregate CID - commP of commPs
+  ferry_link TEXT REFERENCES ferry(link),
+  -- Failed to add into aggregate code
+  ferry_failed_code TEXT,
+);
+
+CREATE INDEX cargo_stat_idx ON cargo (stat);
+CREATE INDEX cargo_car_link_idx ON cargo (car_link);
+CREATE INDEX cargo_aggregate_link_idx ON cargo (ferrylink);
```

Comment:
Thanks, this is great. I like the iteration based on what we talked yesterday! Ended up with a few changes

Added bucket_name and bucket_endpoint to enable us to have information to generate presigned URLs (if in the future we move out of carpark-prod-0 bucket):

```sql
CREATE TABLE content
(
  -- CAR CID
  link TEXT PRIMARY KEY,
  -- CAR Size
  size number NOT NULL,
  -- Bucket name where the CAR is
  bucket_name TEXT NOT NULL,
  -- Endpoint of the bucket
  bucket_endpoint TEXT NOT NULL,
  -- Timestamp
  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
);
```

added timestamp for signed and processed rather than updated so that we can keep better track of the status mutations. There would also be option to have a dealSigned table, but given we will not have any processor for it, I think it is good as is

```sql
CREATE TABLE deal (
  aggregate TEXT PRIMARY KEY NOT NULL REFERENCES aggregate(link),
  status DEAL_STATUS DEFAULT 'PENDING',
  -- if status is an error this may contain details about the error e.g. json containing
  -- piece CIDs that were invalid.
  detail TEXT,
  -- Time when aggregate was send to spade-proxy
  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
  -- Time when aggregate was signed
  signed TIMESTAMP WITH TIME ZONE,
  -- Time when aggregate was processed
  processed TIMESTAMP WITH TIME ZONE
);
```

and finally, added a new table. We were not tracking failure reason. It can also be fetched from receipt, but for that at this level we would need to know the invocation CID, workflow and so on.
-- State for the inclusions that made deals fail
CREATE TABLE inclusion_failure {
  -- Piece CID
  piece TEXT NOT NULL REFERENCES piece(link),
  -- Aggregate CID
  aggregate TEXT NOT NULL REFERENCES aggregate(link),
  -- Failure reason
  reason TEXT NOT NULL,
  -- Time when failure was reported
  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,

  PRIMARY KEY (aggregate, piece)
}
Other option would be to add reason in inclusion table, given it is the only new information. Any thoughts?

On the signing part of the flow, I decided to create https://github.com/web3-storage/w3filecoin/issues/29 to get that in a new iteration once we get rest of the pipeline in place